File: src/main/java/com/daniel/goncharov/algorithm/playground/interviewbit/dp/LongestPalindromicSubsequence.java
package com.daniel.goncharov.algorithm.playground.interviewbit.dp;

public class LongestPalindromicSubsequence {

    public int solve(String inputString) {
        String a = inputString;
        String b = new StringBuilder(inputString).reverse().toString();
        int[][] table = new int[a.length() + 1][b.length() + 1];
        for (int i = 0; i < table.length; i++) {
            int row[] = table[i];
            for (int j = 0; j < row.length; j++) {
                if (i == 0 || j == 0) {
                    row[j] = 0;
                    continue;
                }
                if (a.charAt(i - 1) == b.charAt(j - 1)) {
                    row[j] = table[i - 1][j - 1] + 1;
                    continue;
                } else {
                    row[j] = Math.max(table[i - 1][j], table[i][j - 1]);
                }
            }
        }
        return table[a.length()][b.length()];
    }
}

File: src/main/java/com/daniel/goncharov/algorithm/playground/leetcode/dp/LongestPalindromicSubsequence.kt
package com.daniel.goncharov.algorithm.playground.leetcode.dp

import kotlin.math.max

class LongestPalindromicSubsequence {

    fun longestPalindromeSubseq(seq: String): Int {
        val cache = MutableList(seq.length) { row ->
            MutableList(seq.length) { column ->
                if (row == column) 1
                else 0
            }
        }
        for (startColumn in 1 until seq.length) {
            for (row in 0 until seq.length - startColumn) {
                val column = startColumn + row
                val left = seq[row]
                val right = seq[column]
                when {
                    startColumn == 1 && left == right -> cache[row][column] = 2
                    left == right -> cache[row][column] = 2 + cache[row + 1][column - 1]
                    else -> {
                        cache[row][column] = max(
                            cache[row][column - 1],
                            cache[row + 1][column]
                        )
                    }
                }

            }
        }
        return cache[0][seq.length - 1]
    }

}